src/rgw/rgw_lua.cc

```diff
@@ -175,42 +190,106 @@ int install_packages(const DoutPrefixProvider *dpp, rgw::sal::Driver* driver,
+  if (!tmp_luarocks_path) { 
+    const auto rc = -errno;
+    ldpp_dout(dpp, 1) << "Lua ERROR: failed to create temporary directory from template: " << 
+      tmp_path_template << ". error: " << rc << dendl;
+    return rc;
+  }
+
   // the lua rocks install dir will be created by luarocks the first time it is called
   for (const auto& package : packages) {
     bp::ipstream is;
-    const auto cmd = p.string() + " install --lua-version " + CEPH_LUA_VERSION + " --tree " + luarocks_path + " --deps-mode one " + package;
+    const auto cmd = p.string() + " install --lua-version " + CEPH_LUA_VERSION + " --tree " + tmp_luarocks_path + " --deps-mode one " + package;
     bp::child c(cmd, bp::std_in.close(), (bp::std_err & bp::std_out) > is);
 
     // once package reload is supported, code should yield when reading output
-    std::string line = std::string("CMD: ") + cmd;
+    std::string lines = std::string("Lua CMD: ") + cmd;
+    std::string line;
 
     do {
       if (!line.empty()) {
-        output.append(line);
-        output.append("\n");
+        lines.append("\n\t");
+        lines.append(line);
       }
     } while (c.running() && std::getline(is, line));
 
     c.wait();
     if (c.exit_code()) {
       failed_packages.insert(package);
     }
+    ldpp_dout(dpp, 20) << lines << dendl;
+  }
+  
+  // luarocks directory cleanup
+  std::error_code ec;
+  if (fs::remove_all(luarocks_path, ec)
+      == static_cast<std::uintmax_t>(-1) &&
+      ec != std::errc::no_such_file_or_directory) {
+    ldpp_dout(dpp, 1) << "Lua ERROR: failed to clear luarocks directory: " <<
+      luarocks_path << ". error: " << ec.message() << dendl; 
+    return -ec.value();
+  }
+  auto rc = create_directory_p(dpp, luarocks_path);
+  if (rc < 0) {
+    ldpp_dout(dpp, 1) << "Lua ERROR: failed to recreate luarocks directory: " <<
+      luarocks_path << ". error: " << rc << dendl; 
+    return rc;
+  }
+
+  // switch temporary install directory to luarocks one
+  fs::rename(tmp_luarocks_path, luarocks_path, ec);
+  if (ec) {
+    ldpp_dout(dpp, 1) << "Lua ERROR: failed to switch between temp directory: " <<
+      tmp_luarocks_path << " and luarocks directory: " << luarocks_path << 
+      " . error: " << ec.message() << dendl; 
+    return -ec.value();
   }
 
   return 0;
 }
 
-#endif
+int reload_packages(const DoutPrefixProvider *dpp, rgw::sal::Driver* driver, optional_yield y)
+{
+  auto lua_mgr = driver->get_lua_manager();
+
+  return lua_mgr->reload_packages(dpp, y);
+}
+
+void handle_reload_notify(rgw::sal::LuaManager* lua_manager, optional_yield y, uint64_t notify_id, uint64_t cookie) {
+  if (cookie != lua_manager->watch_handle()) {
+    return;
+  }
+
+  packages_t failed_packages;
+  auto r = install_packages(lua_manager->dpp(), lua_manager->driver(), 
+      y, lua_manager->luarocks_path(), failed_packages);
+  if (r < 0) {
+    ldpp_dout(lua_manager->dpp(), 1) << "WARNING: failed to install lua packages from allowlist. error: " << r
```

Comment:
`WARNING` and `errror` seem at odds.  Perhaps  `WARNING: failed to install Lua packages from allowlist: " << r << dendl;` ?

Reply:
it says "WARNING" because failing to install packages does not necesarily impact service
only if there is a script that want to use these packages the service may be impacted.
mayebe will change to "error code" istead of "error"?

Reply:
maybe "return code"?